reader_compare.go

```diff
@@ -0,0 +1,419 @@
+				return fmt.Errorf("expected readers to have matching big int values")
+			}
+		}
+		break
+	case ion.FloatType:
+		if !isNull {
+			v1, err := reader1.FloatValue()
+			if err != nil {
+				return err
+			}
+
+			v2, err := reader2.FloatValue()
+			if err != nil {
+				return err
+			}
+
+			if math.IsNaN(v1) && math.IsNaN(v2) {
+				if v1 != v2 {
+					return fmt.Errorf("expected readers to have matching float values")
+				}
+			} else if math.IsNaN(v1) || math.IsNaN(v2) {
+				if v1 == v2 {
+					return fmt.Errorf("expected readers to have different float values")
+				}
+			} else if v1 != v2 {
+				return fmt.Errorf("expected readers to have matching float values")
+			}
+		}
+		break
+	case ion.DecimalType:
+		if !isNull {
+			decimal1, err := reader1.DecimalValue()
+			if err != nil {
+				return err
+			}
+
+			decimal2, err := reader2.DecimalValue()
+			if err != nil {
+				return err
+			}
+
+			err = CheckPreciselyEquals(decimal1, decimal2)
+			if err != nil {
+				return err
+			}
+		}
+		break
+	case ion.TimestampType:
+		if !isNull {
+			timestamp1, err := reader1.TimeValue()
+			if err != nil {
+				return err
+			}
+
+			timestamp2, err := reader2.TimeValue()
+			if err != nil {
+				return err
+			}
+
+			if timestamp1 != timestamp2 {
+				return fmt.Errorf("expected readers to have matching timestamp values")
+			}
+		}
+		break
+	case ion.StringType:
+		value1, err := reader1.StringValue()
+		if err != nil {
+			return err
+		}
+
+		value2, err := reader2.StringValue()
+		if err != nil {
+			return err
+		}
+
+		if value1 != value2 {
+			return fmt.Errorf("expected readers to have matching string values")
+		}
+		break
+	case ion.SymbolType:
+		// TODO: Rework this once SymbolTokens are available
+		/*token1 := reader1.SymbolValue()
+		token2 := reader2.SymbolValue()
+
+		if isNull {
+			if token1.Text != nil {
+				return fmt.Errorf("expected token to be null")
+			}
+			if token2.Text != nil {
+				return fmt.Errorf("expected token to be null")
+			}
+		} else if token1.Text == nil || token2.Text == nil {
+			if token1.Sid != token2.Sid {
+				return fmt.Errorf("expected token SIDs to match")
+			}
+		} else if token1.Text != token2.Text {
+			return fmt.Errorf("expected tokens to match")
+		}*/
+		break
+	case ion.BlobType, ion.ClobType:
+		if !isNull {
+			b1, err := reader1.ByteValue()
+			if err != nil {
+				return err
+			}
+
+			b2, err := reader1.ByteValue()
+			if err != nil {
+				return err
+			}
+
+			if b1 == nil || b2 == nil {
+				return fmt.Errorf("expected byte arrays to be not null")
+			}
+
+			if len(b1) != len(b2) {
+				return fmt.Errorf("expected byte arrays to have same length")
+			}
+
+			for i := 0; i < len(b1); i++ {
+				if b1[i] != b2[i] {
+					return fmt.Errorf("expected byte arrays to have matching values")
+				}
+			}
+		}
+		break
+	default:
+		return &InvalidIonTypeError{ionType}
+	}
+
+	return nil
+}
+
+func CheckPreciselyEquals(decimal1, decimal2 *ion.Decimal) error {
```

Comment:
Can we add a comment here explaining what this function does. The name is misleading on a first read.

Reply:
> Compare two Ion Decimal values by equality and negative zero.

Reply:
Thanks. Why is negative zero clarified in that sentence? 
Given that Go does not have overriding functions, perhaps naming this `DecimalStrictEquals` (edit as needed for Go naming conventions) would be better?

Reply:
Done.